File: pinus-ws/tools/FileUtil.ts
import { promises as fs } from 'fs';
import * as crypto from 'crypto';
import * as path from "path";

export async function fileMd5(file_path: string) {
    const content = await fs.readFile(file_path, { encoding: 'utf8' });
    return crypto.createHash('md5').update(content).digest("hex");
}

export async function fileExist(file_path: string): Promise<boolean> {
    let exist = false;
    try {
        await fs.access(file_path);
        exist = true;
    } catch (error) {
    }
    return exist;
}

export async function fileEqual(file_path1: string, file_path2: string): Promise<boolean> {
    const exist1 = await fileExist(file_path1);
    const exist2 = await fileExist(file_path2);
    if (exist1 != exist2) return false;
    if (!exist1 && !exist2) return true;

    const md5_1 = await fileMd5(file_path1);
    const md5_2 = await fileMd5(file_path2);
    return md5_1 == md5_2;
}

export async function walkDir(root: string, ext: string | string[] = [], relative: string = ''): Promise<string[]> {
    // ext -> string[] like [".js", ".css"] or any other extensions
    if (!ext) ext = [];
    if (typeof ext == 'string') ext = [ext];

    const full = path.join(root, relative);
    const files: string[] = [];
    const dir_content = await fs.readdir(full);
    for (let i = 0; i < dir_content.length; i++) {
        const f = dir_content[i];
        const file_path = path.join(full, f);
        const stat = await fs.stat(file_path);

        if (stat?.isDirectory()) {
            const sub = path.join(relative, f);
            const sub_files = await walkDir(root, ext, sub);
            files.push(...sub_files);
        }
        else {
            if (ext.length == 0 || ext.includes(path.extname(f)))
                files.push(path.join(relative, f));
        }
    }
    return files;
}

export async function relativeCopy(
    from: string,
    to: string,
    extensions: string[], // extensions [".js", ".css"] or any other extensions
    skip_check = false) {

    if (!from || !to) {
        throw new Error('invalid arguments');
    }

    if (!await fileExist(from)) {
        throw new Error('from not exist: ' + from);
    }

    const stat_from = await fs.stat(from);
    if (!stat_from?.isDirectory()) {
        // copy file

        if (skip_check || !await fileEqual(from, to)) {
            console.log('copy file: ' + from + ' -> ' + to);
            await fs.mkdir(path.dirname(to), { recursive: true });
            await fs.copyFile(from, to);
        }
        else {
            console.log('skip file: ' + from);
        }

        return;
    }

    // copy folder
    console.log('copy files from: ' + from);
    console.log('\tto: ' + to);
    console.log('\textensions: ' + JSON.stringify(extensions));
    console.log('');

    await fs.mkdir(to, { recursive: true });

    let file_total = 0;
    let file_copied = 0;

    let dir_content = await walkDir(from);

    for (let i = 0; i < dir_content.length; i++) {
        const relative = dir_content[i];
        const ext = path.extname(relative);

        // console.log('\tcheck file: ' + relative);
        if (!extensions || extensions.length <= 0 || extensions.indexOf(ext) != -1) {
            const src = path.join(from, relative);
            const des = path.join(to, relative);

            if (skip_check || !await fileEqual(src, des)) {
                console.log('\tcopied file: ' + relative);
                await fs.mkdir(path.dirname(des), { recursive: true });
                await fs.copyFile(src, des);
                file_copied++;
            }
            else {
                // console.log('build. skip copy equal config file: ' + relative);
            }
            file_total++;
        }
    }

    console.log('total:  \t' + file_total + '\tcopied: \t' + file_copied + '\tskipped:\t' + (file_total - file_copied));
    console.log('copy done.');
    console.log('');
}

File: rust-ws/route-generator/src/utils/file_util.rs
use anyhow::{Ok, Result};
use std::fs;
use std::path::Path;

#[allow(dead_code)]
pub fn file_md5(file_path: &str) -> Result<String> {
    let file = Path::new(file_path);
    let hash = sha256::try_digest(file).unwrap();
    Ok(hash)
}

#[allow(dead_code)]
pub fn file_exist(file_path: &str) -> bool {
    std::path::Path::new(file_path).exists()
}

#[allow(dead_code)]
pub fn file_equal(file_path1: &str, file_path2: &str) -> Result<bool> {
    let exist1 = file_exist(file_path1);
    let exist2 = file_exist(file_path2);
    if exist1 != exist2 {
        return Ok(false);
    }
    if !exist1 && !exist2 {
        return Ok(true);
    }

    let md5_1 = file_md5(file_path1).unwrap();
    let md5_2 = file_md5(file_path2).unwrap();
    Ok(md5_1 == md5_2)
}

#[allow(dead_code)]
pub fn walk_dir(
    root_path_str: &str,
    extensions: &Vec<&str>,
    relative_path_str: &str,
) -> Result<Vec<String>> {
    // find all files in ../../src/handlers
    let mut files: Vec<String> = vec![];
    let dir = Path::new(root_path_str).join(relative_path_str);
    let dir_content = fs::read_dir(dir)
        .unwrap()
        .map(|e| e.unwrap().path())
        .collect::<Vec<_>>();

    for entry_path in dir_content {
        let metadata = fs::metadata(&entry_path).unwrap();
        if metadata.is_dir() {
            walk_dir(root_path_str, extensions, entry_path.to_str().unwrap())?
                .iter()
                .for_each(|f| files.push(f.to_string()));
        } else {
            let mut should_add = false;

            if extensions.len() <= 0 {
                should_add = true;
            } else {
                let ext = Path::extension(entry_path.as_path())
                    .unwrap()
                    .to_str()
                    .unwrap();
                if extensions.contains(&ext) {
                    should_add = true;
                }
            }

            if should_add {
                files.push(String::from(
                    entry_path
                        .strip_prefix(root_path_str)
                        .unwrap()
                        .to_str()
                        .unwrap(),
                ));
            }
        }
    }
    Ok(files)
}

#[allow(dead_code)]
pub fn relative_copy(from: &str, to: &str, extensions: &Vec<&str>, skip_check: bool) -> Result<()> {
    if !file_exist(from) {
        panic!("from not exist: {}", from);
    }

    let stat_from = fs::metadata(from).unwrap();
    if !stat_from.is_dir() {
        // copy file

        if skip_check || !file_equal(from, to).unwrap() {
            println!("copy file: {} -> {}", from, to);
            fs::create_dir_all(Path::new(to).parent().unwrap()).unwrap();
            fs::copy(from, to).unwrap();
        } else {
            println!("skip file: {}", from);
        }

        return Ok(());
    }

    // copy folder
    println!("copy files from: {}", from);
    println!("\tto: {}", to);
    println!("\textensions: {}", extensions.join(","));
    println!("");

    fs::create_dir_all(to).unwrap();

    let mut file_total = 0;
    let mut file_copied = 0;

    let dir_content = walk_dir(from, extensions, "").unwrap();

    for relative in dir_content {
        let relative = Path::new(relative.as_str());
        let src = Path::new(from).join(relative);
        let des = Path::new(to).join(relative);

        if skip_check || !file_equal(src.to_str().unwrap(), des.to_str().unwrap()).unwrap() {
            println!("\tcopied file: {}", relative.to_str().unwrap());
            fs::create_dir_all(des.parent().unwrap()).unwrap();
            fs::copy(src, des).unwrap();
            file_copied += 1;
        } else {
            // println!("build. skip copy equal config file: {}", relative);
        }
        file_total += 1;
    }

    println!(
        "total:  \t{}\tcopied: \t{}\tskipped:\t{}",
        file_total,
        file_copied,
        file_total - file_copied
    );
    println!("copy done.");
    println!("");
    Ok(())
}
